### ISSUE TITLE
Retpoline support in `rustc` (target features or dedicated flags)

### ISSUE THREAD

**Issue by ojeda** [rust-lang/rust#116852]
Currently:

```sh
-Ctarget-feature="+retpoline-external-thunk"
-Ctarget-feature="+retpoline-indirect-branches"
-Ctarget-feature="+retpoline-indirect-calls"
```

seems to work as expected, but the features are unknown, and so since 1.61 they emit a warning:

```txt
warning: unknown feature specified for `-Ctarget-feature`: `retpoline-external-thunk`
  |
  = note: it is still passed through to the codegen backend
  = help: consider filing a feature request

warning: unknown feature specified for `-Ctarget-feature`: `retpoline-indirect-branches`
  |
  = note: it is still passed through to the codegen backend
  = help: consider filing a feature request

warning: unknown feature specified for `-Ctarget-feature`: `retpoline-indirect-calls`
  |
  = note: it is still passed through to the codegen backend
  = help: consider filing a feature request
```

Could/should these be added as known features to avoid using the target specification file?

In an old issue at https://github.com/rust-lang/rust/issues/54637 it seems that back then dedicated flags were not intended (like Clang's `-mretpoline-external-thunk` and GCC's `-mindirect-branch=thunk-extern -mindirect-branch-register`), which perhaps make a bit more sense.

**Comment by RalfJung** [comment]
If I understand @Darksonn correctly in [this discussion](https://github.com/rust-embedded/wg/issues/792), then these target features affect ABI and therefore they should *not* be added to the known feature list. Actually they should be added to the "forbidden features" list that will hopefully be introduced soon (https://github.com/rust-lang/rust/pull/129884), because ABI differences make it unsound to combine code built with and without these target features.

**Comment by ojeda** [comment]
We still need a stable way to disable/enable them (globally). We can use the target specification for the moment, but that is unstable ("Target specification file or enough command-line flags to replace it." in https://github.com/Rust-for-Linux/linux/issues/2).

**Comment by RalfJung** [comment]
Yeah there is clearly a need for "ABI incompatible target variants". Also see <https://github.com/rust-embedded/wg/issues/792>.

**Comment by RalfJung** [comment]
> but that is unstable

Note that soundly adding something like "ABI vairants" requires `-Zbuild-std`, so to get retpoline on stable we need either stable `build-std`, or a separate target.

**Comment by Darksonn** [comment]
I've investigated this issue in more details. It turns out that the retpoline mitigations do *not* affect ABI and can be mixed.

Mixing them gives up their advantage, though. Linking together two compilation units that disagree on whether to apply the mitigation may make *both* vulnerable to what they attempt to mitigate.

**Comment by Darksonn** [comment]
In the kernel Makefile [there is an additional flag `-mindirect-branch-cs-prefix` in `RETPOLINE_CFLAGS`][1]. Do we need that flag too?

[1]: https://github.com/torvalds/linux/blob/b4432656b36e5cc1d50a1f2dc15357543add530e/arch/x86/Makefile#L23
